src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/CqlRetryPolicy.java

```diff
@@ -0,0 +1,179 @@
+    var retryDecision = retryVerdict.getRetryDecision();
+
+    if (LOG.isInfoEnabled()) {
+      LOG.info(
+          RETRYING_ON_READ_TIMEOUT,
+          logPrefix,
+          cl,
+          blockFor,
+          received,
+          false,
+          retryCount,
+          retryDecision);
+    }
+
+    return retryVerdict;
+  }
+
+  @Override
+  public RetryVerdict onWriteTimeoutVerdict(
+      @NonNull Request request,
+      @NonNull ConsistencyLevel cl,
+      @NonNull WriteType writeType,
+      int blockFor,
+      int received,
+      int retryCount) {
+    final RetryDecision retryDecision;
+    // TODO(Hazel): only retry two write type or all?
+    if (retryCount < MAX_RETRIES && (writeType == WriteType.CAS || writeType == WriteType.SIMPLE)) {
+      retryDecision = RetryDecision.RETRY_SAME;
+    } else {
+      retryDecision = RetryDecision.RETHROW;
+    }
+    if (LOG.isInfoEnabled()) {
+      LOG.info(
+          RETRYING_ON_WRITE_TIMEOUT,
+          logPrefix,
+          cl,
+          writeType,
+          blockFor,
+          received,
+          retryCount,
+          retryDecision);
+    }
+    return () -> retryDecision;
+  }
+
+  @Override
+  public RetryVerdict onUnavailableVerdict(
+      @NonNull Request request,
+      @NonNull ConsistencyLevel cl,
+      int required,
+      int alive,
+      int retryCount) {
+    // TODO(Hazel): no error passed in this method, cannot tailor the retry decision for
+    // UnavailableException, override the default retry 1 to retry 3?
+    var retryVerdict = super.onUnavailableVerdict(request, cl, required, alive, retryCount);
+    var retryDecision = retryVerdict.getRetryDecision();
+
+    if (LOG.isInfoEnabled()) {
+      LOG.info(RETRYING_ON_UNAVAILABLE, logPrefix, cl, required, alive, retryCount, retryDecision);
+    }
+
+    return retryVerdict;
+  }
+
+  @Override
+  public RetryVerdict onRequestAbortedVerdict(
+      @NonNull Request request, @NonNull Throwable error, int retryCount) {
+    var retryVerdict = super.onRequestAbortedVerdict(request, error, retryCount);
+    var retryDecision = retryVerdict.getRetryDecision();
+
+    if (LOG.isInfoEnabled()) {
+      LOG.info(RETRYING_ON_ABORTED, logPrefix, retryCount, error, retryDecision);
+    }
+
+    return retryVerdict;
+  }
+
+  @Override
+  public RetryVerdict onErrorResponseVerdict(
+      @NonNull Request request, @NonNull CoordinatorException error, int retryCount) {
+
+    var retryDecision =
+        switch (error) {
+            // TODO(Hazel): what decision? RETRY_SAME or RETRY_NEXT? default doesn't care the
+            // retryCount
+          case CASWriteUnknownException e ->
+              (retryCount < MAX_RETRIES) ? RetryDecision.RETRY_NEXT : RetryDecision.RETHROW;
+
+          case TruncateException e ->
+              (retryCount < MAX_RETRIES) ? RetryDecision.RETRY_NEXT : RetryDecision.RETHROW;
+
+          default -> super.onErrorResponseVerdict(request, error, retryCount).getRetryDecision();
```

Comment:
pull the logic from super class into here and do not call the super because it uses retry_next